transactionoutbox-core/src/main/java/com/gruelbox/transactionoutbox/TransactionOutboxImpl.java

```diff
@@ -146,8 +148,15 @@ private List<TransactionOutboxEntry> flush(Instant now) {
               return result;
             });
     log.debug("Got batch of {}", batch.size());
-    batch.forEach(this::submitNow);
+    batch.stream().filter(entry -> entry.getGroupId() == null).forEach(this::submitNow);
     log.debug("Submitted batch");
+    List<CompletableFuture<Void>> orderedFutures =
+        batch.stream()
+            .filter(entry -> entry.getGroupId() != null)
+            .map(entry -> CompletableFuture.runAsync(() -> this.processNow(entry)))
```

Comment:
Several issues I can see here:

1. We want to make sure that ordered messages never get `blocked` automatically, because that would cause the next message in the group to be processed out of order. Blocking should only ever happen manually if the application can be confident it is safe to do so.  So we need to tweak that behaviour but also make sure there is a well-defined mechanism for blocking an arbitrary entry, e.g. `outbox.block(String id)` and `outbox.unblock(String id`).  Your code currently automatically blocks messages.
2. `CompletableFuture.runAsync` uses the common fork/join pool, which might not be how an application is managing its threads, or might cause the library to block core processing threads which the application isn't expecting.
3. Error handling scenarios are quite different between unordered and ordered. For unordered, we don't really care about errors from `flush()`. If any jobs fail and get blocked, we get errors logged and notifications we can act on, but we basically don't care until then. With ordered flushing, if any work fails, we need to alert immediately; this is a serious situation which is blocking processing.
5.  Also, _all_ the outstanding work gets processed when doing unordered processing, so the caller only needs to call once, whereas with ordered flushing, you will want to call repeatedly until all the work has been processed.

All in all I think the structure here needs a bit of a rethink; I'll post a reply in a moment with some ideas.